src/coreComponents/constitutive/capillaryPressure/BrooksCoreyCapillaryPressure.hpp

```diff
@@ -146,22 +132,13 @@ class BrooksCoreyCapillaryPressure : public CapillaryPressureBase
 
 
 GEOSX_HOST_DEVICE
-GEOSX_FORCE_INLINE
-void
+inline void
 BrooksCoreyCapillaryPressureUpdate::
-  compute( arraySlice1d< real64 const > const & phaseVolFraction,
-           arraySlice1d< real64 > const & phaseCapPres,
-           arraySlice2d< real64 > const & dPhaseCapPres_dPhaseVolFrac ) const
+  compute( arraySlice1d< real64 const, compflow::USD_PHASE_PROP-1 > const & phaseVolFraction,
+           arraySlice1d< real64, cappres::USD_CAPPRES-2 > const & phaseCapPres,
+           arraySlice2d< real64, cappres::USD_CAPPRES_DS-2 > const & dPhaseCapPres_dPhaseVolFrac ) const
```

Comment:
This would be a parameter type mismatch, so there should be a compilation error in at least one of the builds (either non-CUDA or CUDA, depending on type of mistake). With that said, there are some layout/USD parameters that are identical in value, and there is currently no way to prevent misuse.

A similar type of issue comes up when using `getReference<>` with the wrong layout parameter. If you pass the resulting view into a function that expects the correct layout, it's still a compile-time error, but if you use it directly in the same scope (or you wrote the function with the wrong parameter type), it will only manifest itself at runtime. This can be somewhat remediated by using `extrinsicMeshData` pattern, which associates array type (including layout) with the lookup key. I was tempted to switch all compositional variables to that pattern, but eventually decided in favor of a separate PR.

Reply:
By the way, I'm looking for feedback/criticism regarding the choice of different layouts, as well as naming. 
 * Is it too verbose, or not verbose enough? 
 * Do we need this many layout types, e.g. should be merge `PHASE_PROP` and `COMP_PROP` into one layout (I think there's some value in being able to change them independently)?
 * Is  it clear what `-1` and `-2` mean in slice parameters, or should be add explicit names for these values (e.g. `USD_PHASE_PROP-1` -> `USD_PHASE_PROP_ONE_CELL` or something)?
 * Is the namespace choice appropriate? I deviated a bit from the approach used in mechanics - using "physics" to define namespaces (`compflow`) instead of topology (`nodes` and `cells`).

If you have any suggestions on any of that, let me know!

Reply:
- Regarding the level of verbosity, it is fine to me, but one suggestion to trim the layout could be to remove the `PROP`. Since you already have `USD_PHASE_COMP`,  the layout names for phase, components, and fluid could simply be `USD_PHASE`, `USD_COMP`, and `USD_FLUID`, which would be clear, but slightly shorter, and more consistent with `USD_PHASE_COMP`. Regarding the use of `_DC` and `_DS` for the arrays of derivatives, I think it is already clear and concise. Regarding the use of `FLUID` for the arrays storing the quantities that have just one value per quadrature point, it may not be immediately obvious that it means "one value per quadrature point", but it is quite difficult to find something meaningful and concise to replace `FLUID`. I tried to find something better, but I failed ;) 
- Regarding the number of layouts, it is fine with me and I like the distinction between `PHASE` and `COMP`.
- Regarding the slices, that is a difficult one. The downside with the current slice parameters `-1` and `-2` is it forces us to include some information about the structure of the original array, even if the developer of constitutive model is going to manipulate slices that "look" the same whether they come from the solver or the constitutive model. I prefer the `USD_PHASE_PROP_ONE_CELL` or even `USD_PHASE_ONE_CELL`, or `USD_PHASE_LOC` if that can be used for both `compflow` and `multifluid` to hide the `-1` and `-2`. That would be really nice.
- For your namespace choices, I think they make a lot a sense for the compflow solver.